elm/src/Ladder.elm

```diff
@@ -0,0 +1,145 @@
+
+        {- Note the directions and the assymetries between going up and going down:
+           foldl vs foldr,
+           the order that approachingLocation and stopLocation are accumulated,
+           that the extra approaching segment is at the head of the list so it can be dropped,
+           that only the upward list needs be reversed.
+        -}
+        downwardLocations : List Location
+        downwardLocations =
+            stops
+                |> List.foldr
+                    (\stop segmentsSoFar ->
+                        let
+                            stopLocation =
+                                { direction = Downward
+                                , stationId = stop.id
+                                , currentStatus = StoppedAt
+                                }
+
+                            approachingLocation =
+                                { direction = Downward
+                                , stationId = stop.id
+                                , currentStatus = InTransitTo
+                                }
+                        in
+                        approachingLocation :: stopLocation :: segmentsSoFar
+                    )
+                    []
+                -- Drop the segment before the start of the line.
+                |> List.drop 1
+
+        upwardLocations : List Location
+        upwardLocations =
+            stops
+                |> List.foldl
+                    (\stop segmentsSoFar ->
+                        let
+                            stopLocation =
+                                { direction = Upward
+                                , stationId = stop.id
+                                , currentStatus = StoppedAt
+                                }
+
+                            approachingLocation =
+                                { direction = Upward
+                                , stationId = stop.id
+                                , currentStatus = InTransitTo
+                                }
+                        in
+                        approachingLocation :: stopLocation :: segmentsSoFar
+                    )
+                    []
+                -- Drop the segment before the start of the line.
+                |> List.drop 1
+                |> List.reverse
+
+        stopNames : List (Maybe String)
+        stopNames =
+            stops
+                |> List.map .name
+                |> List.map Just
+                |> List.intersperse Nothing
+    in
+    List.map3
+        (\downLocation upLocation stopName ->
+            { leftTrain =
+                vehiclesByLocation
+                    |> Dict.member downLocation
+            , rightTrain =
+                vehiclesByLocation
+                    |> Dict.member upLocation
+            , stopName = stopName
+            }
+        )
+        downwardLocations
+        upwardLocations
+        stopNames
+
+
+groupBy : (v -> k) -> List v -> Dict.Dict k (List v)
```

Comment:
This is awesome. Of this changeset, I think groupBy is the one part that I had the most trouble understanding off the bat. Could this be done in less code with a sort?

Reply:
Thanks for the explanation. I was indeed referring to making the dict with a sort.

Knowing now the background behind your custom groupBy, I'm more okay with it. I like the idea of using `List.filter`, as that does seem like it could be a drop-in, simpler replacement. I'm indifferent though; I'd feel comfortable with either being merged now that I'm clued in.

Reply:
👍